add BLOB type

mattn opened the pull request:
This change add new type v:t_BLOB to handle bytes contains NUL byte. Current implementation have following features. To know more description, please read my slide(P87-P102) of [keynote speach](https://docs.google.com/presentation/d/1ZfGxx_5Wu31rWo_MYtYNDCsl89h54z3L2jRzoFxXzfs/edit#slide=id.g42e6cc335f_0_1) on [VimConf2018](https://vimconf.org/2018/).

I'll add tests in later.

mattn committed:
support blob

mattn committed:
support blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
blob

mattn committed:
Merge branch 'master' into blob

mattn committed:
Merge branch 'master' into blob

mattn committed:
fix index()

mattn committed:
fix blob

mattn committed:
fix blob

mattn committed:
fix blob

mattn committed:
let blob[0:1]

mattn committed:
Merge branch 'master' into blob

mattn committed:
Merge branch 'blob' of bitbucket.org:mattn/vim into blob

chrisbra commented:
This pull request **introduces 2 alerts** when merging e6470d5ba47daa47685cd8db052afcea09e55400 into 3067a4dd0d768d1e4a47cf3c100f3e2b462717d7 - [view on LGTM.com](https://lgtm.com/projects/g/vim/vim/rev/pr-0404d59714d3bdc2851d6565c0d30b73c2d94a88)

**new alerts:**

* 1 for Declaration hides parameter
* 1 for Comparison result is always the same

---

*Comment posted by [LGTM.com](https://lgtm.com)*

mattn commented:
Sorry for my bad commit messages

mattn commented on runtime/doc/eval.txt:
```diff
@@ -57,6 +57,7 @@ List		An ordered sequence of items |List|.
 		Example: [1, 2, ['a', 'b']]
 
 Dictionary	An associative, unordered array: Each entry has a key and a
+		Example: 0zFF00ED015DA
```
will remove

bfredl commented:
It would be nice with initialization similar to string literals, i e something like `b"binary\x00\x03data"` (probably exact that syntax doesn't work, but something in that spirit).

mattn committed:
Merge branch 'blob' of github.com:mattn/vim into blob

mattn committed:
fix blob

mattn committed:
add test

chrisbra commented:
This pull request **introduces 2 alerts** when merging 59c22859fe6b790be76005b6b704d522876ab3f8 into 10efcd5b0259659cd9a152a7a342deb5d56a8eb5 - [view on LGTM.com](https://lgtm.com/projects/g/vim/vim/rev/pr-a432214ae116da28dd8d0398621513d88e76f9b1)

**new alerts:**

* 1 for Declaration hides parameter
* 1 for Comparison result is always the same

---

*Comment posted by [LGTM.com](https://lgtm.com)*

mattn committed:
fix Makefile

mattn committed:
fix Make_mvc.mak

chrisbra commented:
This pull request **introduces 2 alerts** when merging 4ccd5f83294ccce6082a80fb02bb7a8ab84301b1 into 10efcd5b0259659cd9a152a7a342deb5d56a8eb5 - [view on LGTM.com](https://lgtm.com/projects/g/vim/vim/rev/pr-7a8ee3c9a071ce8b67c85c447adc1a68f0732ebf)

**new alerts:**

* 1 for Declaration hides parameter
* 1 for Comparison result is always the same

---

*Comment posted by [LGTM.com](https://lgtm.com)*

mattn committed:
Fix doc

chrisbra commented:
This pull request **introduces 2 alerts** when merging d588615566076727660a9ca268c8475f3252d22d into 10efcd5b0259659cd9a152a7a342deb5d56a8eb5 - [view on LGTM.com](https://lgtm.com/projects/g/vim/vim/rev/pr-1dcb0770aa9a0d960fd655e380dfedc98eaf1067)

**new alerts:**

* 1 for Declaration hides parameter
* 1 for Comparison result is always the same

---

*Comment posted by [LGTM.com](https://lgtm.com)*

mattn committed:
Fix compilation warnings

h-east committed:
I made the following modifications.
- The compiler warning has been removed.
- Fixed blob.c being specified twice in BASIC_SRC in Makefile.
- We added the file addition position in the Makefile as much as possible to the alphabet order.
- Changed comment style of added code to "//".

mattn committed:
Merge pull request #1 from h-east/blob

I made the following modifications.

mattn committed:
support blob in if_python3

mattn committed:
add String#to_blob to return string 0zXXXX in if_ruby

h-east commented:
You should add `test_blob` to `NEW_TESTS` in Make_all.mak for individual test:+1:

mattn committed:
add VIM::Blob() to if_perl

mattn committed:
add test

mattn committed:
add VIM::blob() and remove String.to_blob

mattn committed:
update doc

mattn commented:
Added Vim::blob in if_ruby, VIM::Blob in if_perl. if_python3 can translate blob as bytes.

k-takata commented:
I think `v:t_blob` should be documented in `:help v:t_TYPE` and `:help type()`.

mattn committed:
Merge branch 'master' into blob

mattn committed:
Update doc

k-takata commented on runtime/doc/eval.txt:
```diff
@@ -9044,6 +9059,7 @@ type({expr})	The result is a Number representing the type of {expr}.
 			None	    7  |v:t_none| (v:null and v:none)
 			Job	    8  |v:t_job|
 			Channel	    9  |v:t_channel|
+			Channel	   10  |v:t_blob|
```
Channel -> Blob

k-takata commented on src/eval.c:
```diff
@@ -7113,6 +7385,9 @@ get_tv_string_buf_chk(typval_T *varp, char_u *buf)
 	    STRCPY(buf, get_var_special_name(varp->vval.v_number));
 	    return buf;
+        case VAR_BLOB:
+	    EMSG(_("E964: using Blob as a String"));
```
The wording differ to E963 above.
This should be "E964: Using a Blob as a String".
(E963 has been already used by another patch recently merged. So the error numbers should be updated when merging.)

mattn replied:
Thanks yoru review. There are some wording using lower letter in few lines above.

https://github.com/vim/vim/pull/3638/files/324aca2e5df08652dccecfc3dbaeaa7115341502#diff-de9ab8906fef668ed8042046e0a1eb02R7374

@brammool Can I update them too?

k-takata commented on src/eval.c:
```diff
@@ -10324,6 +10646,39 @@ filter_map(typval_T *argvars, typval_T *rettv, int map)
+		vimvars[VV_KEY].vv_nr = idx;
+		tv.vval.v_number = blob_get(b, i);
+		if (filter_map_one(&tv, expr, map, &rem) == FAIL
+								  || did_emsg)
```
Maybe these 2 lines can be joined.

mattn committed:
Fixes via k-takata's review

mattn committed:
Check bytes strictly

mattn committed:
remove converting PyBytes

mattn commented:
I removed convertion of PyBytes since tests are failed.

brammool commented:
Is this ready to include now?

mattn commented:
Will fix conflict. And I need to check error code.

mattn committed:
Merge branch 'master' into blob

mattn committed:
Merge branch 'blob' of github.com:mattn/vim into blob

mattn commented:
I'm okay for merge.
